java-checks/src/main/java/org/sonar/java/checks/VirtualThreadNotSynchronizedCheck.java

```diff
@@ -0,0 +1,229 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.JavaVersionAwareVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6906")
+public class VirtualThreadNotSynchronizedCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava21Compatible();
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.COMPILATION_UNIT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    var runnablesCollector = new RunnablesToCheckCollector();
+    tree.accept(runnablesCollector);
+
+    var synchonizedTreesCollector = new TreesWithSynchronizedCodeCollector();
+    runnablesCollector.invocations.stream()
+      .filter(it -> synchonizedTreesCollector.isInvokingSynchronizedCode(it.arguments().get(0)))
+      .forEach(it -> reportIssue(it, synchonizedTreesCollector.getSecondaryLocation(it.arguments().get(0))));
+  }
+
+  void reportIssue(MethodInvocationTree tree, Tree secondaryLocation) {
+    reportIssue(
+      tree.methodSelect(),
+      "Use a platform thread instead of a virtual thread",
+      List.of(new JavaFileScannerContext.Location("synchronized", secondaryLocation)),
+      null
+    );
+  }
+
+  private static class RunnablesToCheckCollector extends BaseTreeVisitor {
```

Comment:
Yes: the main check is empty, so it can be stateless. That's a pattern I use when the actual implementation needs a state (like the HashSet and HashMap in my case), since we cannot rely on any cleanup-function called on the check after a file is visited.

`AbstractMethodDetection` I didn't use only because of this. I don't want a state in the actual Check class, but `AbstractMethodDetection` is a `IssuableSubscriptionVisitor`. I looked whether we have a `BaseTreeVisitor` implementation with similar functionality, but found nothing. 

I'm happy to discuss that though.